### PR TITLE
[primTorch] Move addcdiv from decompositions -> refs

### PR DESCRIPTION
**Pull request by nkaretnikov**

### CODE REVIEW

File: torch/_refs/__init__.py
Comment (Lezcano) [inline torch/_refs/__init__.py:1328]: If `value` is a float, why are you assigning an integer to it? Also, isn't `value` a `NumberType`?

File: torch/_refs/__init__.py
Comment (Lezcano) [inline torch/_refs/__init__.py:1329]: We may need some logic here similar to the logic for `alpha` in `aten.add` when it comes to castings.

File: torch/_refs/__init__.py
Comment (Lezcano) [inline torch/_refs/__init__.py:1328]: If value is a float, why are you assigning an integer to it? Also, isn't value a NumberType?

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:20793]: It may be interesting to add extra reference samples that test for:
- Broadcasting
- Casting with respect to `value`

File: torch/_refs/__init__.py
Comment (Lezcano) [reply torch/_refs/__init__.py:1329]: Also, how does this implementation relate to the `warning` in the docs https://pytorch.org/docs/stable/generated/torch.addcdiv.html? Does this implementation work as expected for integer tensors? Could you add some reference inputs to the OpInfo (note that this is not the same as the sample inputs) to test for this?

**Comment by nkaretnikov** [comment]
**please don't merge this yet even if ci passes. i want to address mario's comments first.**

this was converted from a draft because the draft one wasn't triggering ci for some instance on force-pushes

File: torch/testing/_internal/common_methods_invocations.py
Comment (nkaretnikov) [reply torch/testing/_internal/common_methods_invocations.py:20783]: - broadcasting is already covered by sample inputs
- promotion seems broken, so i'll try another attempt to land this one first: https://github.com/pytorch/pytorch/pull/74234

File: torch/_refs/__init__.py
Comment (nkaretnikov) [reply torch/_refs/__init__.py]: this is now addressed as in the other refs: doing it any other way seems to make the linter unhappy, which is why it's None in the type signature and not some number

File: torch/_refs/__init__.py
Comment (Lezcano) [inline torch/_refs/__init__.py:1329]: `addcdiv` does not support numbers, why do you support numbers here?

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:20783]: Land, or simply make the function more restrictive, meaning: Just check that the dtypes of everything are the same in both the C++ implementation and the ref implementation and roll with it.

**Review by lezcano — COMMENTED** [review]
Just two small comments, otherwise looks good.

**Review by lezcano — APPROVED** [review]
Just a couple minor comments, but this looks great! Thank you @nkaretnikov!
@mruberry could you have a look at this one?

File: test/test_type_promotion.py
Comment (nkaretnikov) [reply test/test_type_promotion.py:935]: yes, i think so. i think it's not too costly to have this while it might still catch some things. it can be removed later if it's a problem. if you feel strongly about it, it can be removed when the addcmul ref lands.

Comment (Lezcano) [reply test/test_type_promotion.py:935]: what are the "some things" that this may catch that the OpInfo may not?

Comment (nkaretnikov) [reply test/test_type_promotion.py:935]: it's just that opinfo/sample/ref-inputs have too many moving bits while this is a pretty transparent test for a concrete issue. but i guess you're right, i'm going to remove it

File: torch/testing/_internal/common_methods_invocations.py
Comment (nkaretnikov) [reply torch/testing/_internal/common_methods_invocations.py:3247]: i pasted from the type promotion test initially, then ran the test here and it only complained about the first input. do i understand correctly that this will prevent the graph from building, hence might miss something? i'll add the argument to these two.

Comment (nkaretnikov) [reply torch/testing/_internal/common_methods_invocations.py:3287]: heads up: it's been added to make_arg, so doesn't show up in the inline code view

**Comment by nkaretnikov** [comment]
@Lezcano ptal

the failing test is broken on hud as well, see this commit:
```
11:31 AM | af0160c | [torchdynamo hash update] update the pinned torchdynamo hash (#81156) | #81156 | pytorchmerge
```

**Comment by nkaretnikov** [comment]
@Lezcano as discussed offline, looked at it again and noticed i don't pass value=None in the test, fixed now. not planning to push any changes. ptal. sorry for the noise!

**Review by lezcano — APPROVED** [review]

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1410]: Nice wrapper

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1411]: Is this correct? 

Today `torch.addcdiv` will actually throw an error when given integer tensors because that behavior is deprecated as part of our changes to division in PyTorch, where it went from being C-style division to Python 3.0-like true division. We can actually update the behavior of `torch.addcdiv` in ATen at this point to perform true division and be consistent with the implementation here.

However, if `addcdiv` performs true division then I think it should implement `INT_TO_FLOAT` type promotion to be consistent with `true_divide`?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:10896]: Can `addcdiv` (and `addcmul`) use the ternary inputs that clamp uses? See

```
    OpInfo('clamp',
           aliases=('clip',),
           ref=_clamp_numpy,
           dtypes=all_types_and(torch.bfloat16),
           dtypesIfCUDA=all_types_and(torch.half, torch.bfloat16),
           sample_inputs_func=sample_inputs_clamp,
           reference_inputs_func=partial(reference_inputs_elementwise_ternary, sample_inputs_func=sample_inputs_clamp),
```

**Comment by nkaretnikov** [comment]
@mruberry do i need to report the [xla promotion failure](https://github.com/pytorch/pytorch/runs/7323967829?check_suite_focus=true) to the xla repo and skip this test on xla? or is `promote_integer_inputs_to_float`, which i'm enabling here for `addcdiv`, is supposed to be used with other casting TI options?

right now, i have this for `addcdiv`:

```cpp
void TensorIteratorBase::build_ternary_float_op(
    const TensorBase& out, const TensorBase& a,
    const TensorBase& b, const TensorBase& c) {
  build(TensorIteratorConfig()
      .promote_inputs_to_common_dtype(true)
      .enforce_safe_casting_to_output(true)
      .promote_integer_inputs_to_float(true)
      .add_owned_output(out)
      .add_owned_input(a)
      .add_owned_input(b)
      .add_owned_input(c));
}
```

and `div` uses this, which the failing test is compared against:

```cpp
#define BINARY_FLOAT_OP_CONFIG()                \
  TensorIteratorConfig()                        \
    .set_check_mem_overlap(true)                \
    .allow_cpu_scalars(true)                    \
    .promote_inputs_to_common_dtype(true)       \
    .cast_common_dtype_to_outputs(true)         \
    .enforce_safe_casting_to_output(true)       \
    .promote_integer_inputs_to_float(true)

void TensorIteratorBase::build_borrowing_binary_float_op(
    const TensorBase& out, const TensorBase& a, const TensorBase& b) {
  build(BINARY_FLOAT_OP_CONFIG()
        .add_output(out)
        .add_input(a)
        .add_input(b));
}
```

example error on xla:

```
======================================================================
FAIL [0.009s]: test_div_promotion_xla_uint8 (__main__.TestTypePromotionXLA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/var/lib/jenkins/workspace/xla/test/../../test/test_type_promotion.py", line 36, in wrapped_fn
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/xla/test/../../test/test_type_promotion.py", line 743, in test_div_promotion
    self.assertEqual(casting_result, op(dividend, divisor))
  File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 628, in assertEqual
    return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2334, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: The values for attribute 'dtype' do not match: torch.float32 != torch.uint8.
```

**Comment by nkaretnikov** [comment]
@mruberry ... err, i should have said this: `torch.div` and `addcdiv` behave differently. both are tested against `/`

**Comment by nkaretnikov** [comment]
i reverted the branch to the pre-int-to-float promotion stage and will address the remaining reviews within the scope of this PR. the branch that adds int-to-float promotion (which errors out on xla as described above) to addcdiv lives [here](https://github.com/nkaretnikov/pytorch/commits/primtorch-addcdiv-with-int-to-float-promotion)

**Comment by mruberry** [comment]
> @mruberry do i need to report the [xla promotion failure](https://github.com/pytorch/pytorch/runs/7323967829?check_suite_focus=true) to the xla repo and skip this test on xla? or is `promote_integer_inputs_to_float`, which i'm enabling here for `addcdiv`, is supposed to be used with other casting TI options?
> 
> right now, i have this for `addcdiv`:
> 
> ```c++
> void TensorIteratorBase::build_ternary_float_op(
>     const TensorBase& out, const TensorBase& a,
>     const TensorBase& b, const TensorBase& c) {
>   build(TensorIteratorConfig()
>       .promote_inputs_to_common_dtype(true)
>       .enforce_safe_casting_to_output(true)
>       .promote_integer_inputs_to_float(true)
>       .add_owned_output(out)
>       .add_owned_input(a)
>       .add_owned_input(b)
>       .add_owned_input(c));
> }
> ```
> 
> and `div` uses this, which the failing test is compared against:
> 
> ```c++
> #define BINARY_FLOAT_OP_CONFIG()                \
>   TensorIteratorConfig()                        \
>     .set_check_mem_overlap(true)                \
>     .allow_cpu_scalars(true)                    \
>     .promote_inputs_to_common_dtype(true)       \
>     .cast_common_dtype_to_outputs(true)         \
>     .enforce_safe_casting_to_output(true)       \
>     .promote_integer_inputs_to_float(true)
> 
> void TensorIteratorBase::build_borrowing_binary_float_op(
>     const TensorBase& out, const TensorBase& a, const TensorBase& b) {
>   build(BINARY_FLOAT_OP_CONFIG()
>         .add_output(out)
>         .add_input(a)
>         .add_input(b));
> }
> ```
> 
> example error on xla:
> 
> ```
> ======================================================================
> FAIL [0.009s]: test_div_promotion_xla_uint8 (__main__.TestTypePromotionXLA)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
>     result = test(self, **param_kwargs)
>   File "/var/lib/jenkins/workspace/xla/test/../../test/test_type_promotion.py", line 36, in wrapped_fn
>     fn(*args, **kwargs)
>   File "/var/lib/jenkins/workspace/xla/test/../../test/test_type_promotion.py", line 743, in test_div_promotion
>     self.assertEqual(casting_result, op(dividend, divisor))
>   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 628, in assertEqual
>     return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2334, in assertEqual
>     msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
>     raise error_metas[0].to_error(msg)
> AssertionError: The values for attribute 'dtype' do not match: torch.float32 != torch.uint8.
> ```

I would just skip the test on XLA and report the issue

**Comment by Lezcano** [comment]
After discussing with Nikita, we'll keep this PR simple, as it's very close to being mergeable, and then we'll put together a follow-up PR with the necessary ATen changes to `addcdiv`.

**Comment by nkaretnikov** [comment]
@mruberry ptal. this xla failure seems unrelated (hud is full of xla errors now)

**Comment by Lezcano** [comment]
@ngimel this one should be ready for review. As discussed above, this one implements a forward-looking `addcdiv`, implementing also the behaviour for integers with the correct division. We will submit a PR later on fixing the behaviour of `addcdiv` itself in ATen.

**Comment by nkaretnikov** [comment]
@ngimel would you prefer me to rebase this just in case?

**Comment by ngimel** [comment]
Yes please

File: test/test_type_promotion.py
Comment (ngimel) [inline test/test_type_promotion.py:935]: why is type promotion test removed?

Comment (nkaretnikov) [reply test/test_type_promotion.py:935]: it's handled by the opinfo tests now. see this discussion with mario: https://github.com/pytorch/pytorch/pull/80842#discussion_r915740206

**Comment by nkaretnikov** [comment]
@ngimel ci is green. ptal. NOTE: i had to skip this test after rebasing: https://github.com/pytorch/pytorch/pull/80842/commits/28de17010201e352a187ff5757ec6947db42614e

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@nkaretnikov can you please take a look why did you have to skip the test? 82 seems like a very large difference, are the sample inputs just wrong? Is type promotion doing weird things?

**Comment by nkaretnikov** [comment]
@ngimel i bisected the range of commits before and after rebase, this commit introduced an error for float32 (if you apply patches from this branch, without the skip):

```
 7a4a8f327a Add new NVFuser Python Frontend Record Keeping for Cache enablement. (#81578) 
```

**Comment by ngimel** [comment]
cc @kevinstephano. @nkaretnikov is this error with nvfuser executor? If that's the case, you should just skip nvfuser test (set `.supports_nvfuser=False` on your OpInfo)

**Pull request closed without merging**
